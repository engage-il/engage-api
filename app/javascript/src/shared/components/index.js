// @flow
export { Defer } from './Defer'
export { TranslateAndFade } from './TranslateAndFade'
export { NotificationView } from './NotificationView'

export { Link } from './Link'
export type { LinkProps } from './Link'
export { CopyLink } from './CopyLink'
export { Button } from './Button'
export { CloseButton } from './CloseButton'

export { BillSearch } from './BillSearch'
