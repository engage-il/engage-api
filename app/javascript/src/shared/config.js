// @flow
// set this to your desired env key; try not to check in changes to this value
const debugEnv = null

type Config = {|
  graphUrl: string,
|}

export function loadConfig () {
  const configs: { [key: string]: Config } = {
    development: {
      graphUrl: 'http://localhost:3000/api/graphql'
    },
    staging: {
      graphUrl: 'https://legislated-staging.herokuapp.com/api/graphql'
    },
    production: {
      graphUrl: 'https://legislated.herokuapp.com/api/graphql'
    }
  }

  const buildEnv = process.env.ENVIRONMENT
  const env = buildEnv === 'development' ? debugEnv || buildEnv : buildEnv
  if (!env) {
    throw new Error('No environment specified!')
  }

  const config = configs[env]
  if (!config) {
    throw new Error(`No config for environment: ${env}!`)
  }

  return config
}

export default loadConfig()
