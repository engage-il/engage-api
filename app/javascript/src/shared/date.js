export { default as addDays } from 'date-fns/add_days'
export { default as addHours } from 'date-fns/add_hours'
export { default as addMonths } from 'date-fns/add_months'
export { default as differenceInHours } from 'date-fns/difference_in_hours'
export { default as endOfDay } from 'date-fns/end_of_day'
export { default as formatDate } from 'date-fns/format'
export { default as parseDate } from 'date-fns/parse'
export { default as startOfDay } from 'date-fns/start_of_day'