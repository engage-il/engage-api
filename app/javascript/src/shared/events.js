// @flow
import EventEmitter from 'eventemitter3'

class Events extends EventEmitter {
  setAuthHeader = '@@events/set-auth-header'
  setEnvironment = '@@events/set-environment'
}

export const events = new Events()
