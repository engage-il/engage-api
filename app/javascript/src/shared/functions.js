// @flow
import { omit } from 'lodash'

export function withoutRouter (props: Object) {
  return omit(props, [
    'location',
    'history',
    'match',
    'staticContext'
  ])
}

export function sleep (duration: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, duration))
}
