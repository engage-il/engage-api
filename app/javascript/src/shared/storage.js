/* globals Storage, localStorage, sessionStorage */
// @flow
// keys
type LocalStoreKey
  = 'intro-visited'
  | 'intro-cleared'

type SessionStoreKey
  = 'last-search-count'
  | 'admin-header'

// store type
function namespaceKey (key: string): string {
  return `@@legislated/${key}`
}

class Store<K: string> {
  storage: Storage

  constructor (storage: Storage) {
    this.storage = storage
  }

  get (key: K): ?string {
    const actualKey = namespaceKey(key)
    return this.storage.getItem(actualKey)
  }

  set (key: K, value: ?string) {
    const actualKey = namespaceKey(key)

    if (value) {
      this.storage.setItem(actualKey, value)
    } else {
      this.storage.removeItem(actualKey)
    }
  }
}

// store instances
export const local: Store<LocalStoreKey> =
  new Store(localStorage)

export const session: Store<SessionStoreKey> =
  new Store(sessionStorage)
