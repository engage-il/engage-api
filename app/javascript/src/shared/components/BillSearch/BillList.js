// @flow
import * as React from 'react'
import { createPaginationContainer, graphql } from 'react-relay'
import type { RelayPaginationProp } from 'react-relay'
import { withRouter } from 'react-router-dom'
import type { ContextRouter } from 'react-router-dom'
import { initialVariables } from './BillSearch'
import { BillCell } from './BillCell'
import { LoadMoreButton } from './LoadMoreButton'
import { TranslateAndFade } from 'shared/components'
import { formatDate } from 'shared/date'
import { session } from 'shared/storage'
import { stylesheet, mixins } from 'shared/styles'
import { withLoadMoreArgs } from 'shared/relay'
import type { Viewer } from 'shared/types'

type Props = {
  relay: RelayPaginationProp,
  viewer: Viewer,
  animated: Boolean
} & ContextRouter

type State = {
  disableAnimations: boolean
}

function format (date: Date): string {
  return formatDate(date, 'MMM Do')
}

function formatCount ({ bills }: Viewer) {
  return `Found ${bills.count} result${bills.count === 1 ? '' : 's'}.`
}

let BillList = class BillList extends React.Component<*, Props, State> {
  state = {
    disableAnimations: this.props.history.action === 'POP'
  }

  // events
  didClickLoadMore = () => {
    const { relay } = this.props
    if (!relay.hasMore() || relay.isLoading()) {
      return
    }

    relay.loadMore(initialVariables.count, (error: ?Error) => {
      if (error) {
        console.error(`error loading next page: ${error.toString()}`)
      }
    })
  }

  // lifecycle
  componentDidMount () {
    const { history } = this.props
    if (history.action === 'POP') {
      this.setState({ disableAnimations: false })
    }
  }

  componentWillUnmount () {
    const { viewer } = this.props
    if (viewer) {
      session.set('last-search-count', `${viewer.bills.edges.length}`)
    }
  }

  render () {
    const { disableAnimations } = this.state
    const { relay, viewer, animated } = this.props
    const { startDate, endDate } = initialVariables

    return <div {...rules.container}>
      <div {...rules.header}>
        <h2>Upcoming Bills</h2>
        <div>{`${format(startDate)} to ${format(endDate)}`}</div>
        <div>{formatCount(viewer)}</div>
      </div>
      <TranslateAndFade disable={!animated || disableAnimations}>
        {viewer.bills.edges.map(({ node }) => (
          <BillCell key={node.id} bill={node} />
        ))}
      </TranslateAndFade>
      <LoadMoreButton
        styles={rules.loadMoreButton}
        hasMore={relay.hasMore()}
        onClick={this.didClickLoadMore}
      />
    </div>
  }
}

BillList = createPaginationContainer(withRouter(BillList),
  graphql`
    fragment BillList_viewer on Viewer {
      bills(
        first: $count, after: $cursor,
        query: $query, from: $startDate, to: $endDate
      ) @connection(key: "BillList_bills") {
        count
        pageInfo {
          hasNextPage
          endCursor
        }
        edges {
          node {
            id
            ...BillCell_bill
          }
        }
      }
    }
  `,
  withLoadMoreArgs({
    getConnectionFromProps (props) {
      return props.viewer && props.viewer.bills
    },
    query: graphql`
      query BillListQuery(
        $count: Int!, $cursor: String!,
        $query: String!, $startDate: Time!, $endDate: Time!
      ) {
        viewer {
          ...BillsList_viewer
        }
      }
    `
  })
)

const rules = stylesheet({
  container: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'stretch'
  },
  header: {
    marginBottom: 15,
    '> h2': {
      display: 'inline-block',
      marginBottom: 5,
      ...mixins.mobile.glam({
        marginBottom: 0
      })
    },
    '> div': {
      fontSize: 18,
      ':first-of-type': {
        display: 'inline-block',
        marginLeft: 5
      },
      ...mixins.mobile.glam({
        fontSize: 16,
        ':first-of-type': {
          display: 'none'
        }
      })
    }
  },
  loadMoreButton: {
    alignSelf: 'center',
    marginTop: 30,
    ...mixins.mobile.glam({
      marginTop: 20
    })
  }
})

export { BillList }
