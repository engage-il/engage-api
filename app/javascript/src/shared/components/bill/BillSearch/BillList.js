// @flow
import * as React from 'react'
import { createPaginationContainer, graphql } from 'react-relay'
import type { RelayPaginationProps } from 'react-relay'
import { withRouter } from 'react-router-dom'
import type { ContextRouter } from 'react-router-dom'
import styled from 'react-emotion'
import { BillCell } from './BillCell'
import { BILL_SPACING, BILL_SPACING_MOBILE } from './constants'
import { TranslateAndFade, Button } from '@/components'
import { session } from '@/storage'
import { colors, mixins } from '@/styles'
import type { Viewer } from '@/types'

type Props = {
  viewer: Viewer,
  isAnimated: boolean,
  pageSize?: number
} & ContextRouter & RelayPaginationProps

type State = {
  disablesAnimation: boolean
}

function formatCount ({ bills }: Viewer) {
  return `Found ${bills.count} result${bills.count === 1 ? '' : 's'}.`
}

let BillList = class BillList extends React.Component<Props, State> {
  state = {
    disablesAnimation: this.isFromPop
  }

  // accessors
  get isFromPop () {
    return this.props.history.action === 'POP'
  }

  // events
  didClickLoadMore = () => {
    const { relay, pageSize } = this.props
    if (!relay.hasMore() || relay.isLoading()) {
      return
    }

    pageSize && relay.loadMore(pageSize, (error: ?Error) => {
      if (error) {
        console.error(`error loading next page: ${error.toString()}`)
      }
    })
  }

  // lifecycle
  componentDidMount () {
    if (this.isFromPop) {
      this.setState({ disablesAnimation: false })
    }
  }

  componentWillUnmount () {
    const { viewer } = this.props
    viewer && session.set('last-search-count', `${viewer.bills.edges.length}`)
  }

  render () {
    const {
      viewer,
      isAnimated,
      pageSize,
      relay
    } = this.props

    const {
      disablesAnimation
    } = this.state

    return (
      <Bills>
        <h5>{formatCount(viewer)}</h5>
        <TranslateAndFade
          component={List}
          disable={!isAnimated || disablesAnimation}
        >
          {viewer.bills.edges.map(({ node }) => (
            <BillCell key={node.id} bill={node} />
          ))}
        </TranslateAndFade>
        {pageSize && relay.hasMore() && (
          <ActionButton
            isSecondary
            onClick={this.didClickLoadMore}
            children='Load More'
          />
        )}
      </Bills>
    )
  }
}

BillList = withRouter(createPaginationContainer(BillList, graphql`
  fragment BillList_viewer on Viewer {
    bills(
      params: $params,
      first: $count,
      after: $cursor
    ) @connection(key: "BillList_bills") {
      count
      pageInfo {
        hasNextPage
        endCursor
      }
      edges {
        node {
          id
          ...BillCell_bill
        }
      }
    }
  }
`, {
  direction: 'forward',
  getConnectionFromProps ({ viewer }) {
    return viewer && viewer.bills
  },
  getVariables (props, { count, cursor }, previous) {
    return { ...previous, count, cursor }
  },
  query: graphql`
    query BillListQuery(
      $params: BillsSearchParams!,
      $count: Int!,
      $cursor: String!
    ) {
      viewer {
        ...BillList_viewer
      }
    }
  `
}))

const Bills = styled.div`
  ${mixins.flexColumn};

  > h5 {
    margin-bottom: ${BILL_SPACING}px;

    ${mixins.mobile`
      margin-bottom: ${BILL_SPACING_MOBILE}px;
    `}
  }
`

const List = styled.div`
  > * {
    border-bottom: 1px solid ${colors.gray4};
    margin-bottom: ${BILL_SPACING}px;
    padding-bottom: ${BILL_SPACING}px;

    ${mixins.mobile`
      margin-bottom: ${BILL_SPACING_MOBILE}px;
      padding-bottom: ${BILL_SPACING_MOBILE}px;
    `}
  }
`

const ActionButton = styled(Button)`
  align-self: center;
  margin-bottom: 70px;

  ${mixins.mobile`
    margin-bottom: 50px;
  `}
`

export { BillList }
