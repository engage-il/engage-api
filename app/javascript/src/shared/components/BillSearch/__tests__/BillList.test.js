/* eslint-env jest */
import * as React from 'react'
import { shallow } from 'enzyme'
import { defaultsDeep } from 'lodash'
import { BillList } from '../BillList'
import { session } from 'shared/storage'

const { anything } = expect

// mocks
jest.mock('../BillSearch', () => {
  const { addDays } = require('shared/date')

  const initialVariables = {
    count: 999,
    startDate: new Date(2016, 12, 1),
    endDate: addDays(new Date(2016, 12, 1), 6)
  }

  return { initialVariables }
})

// subject
let subject

const defaults = {
  animated: true,
  viewer: {
    bills: {
      count: 2,
      edges: [{ node: { id: 1 } }, { node: { id: 2 } }]
    }
  },
  history: {
    action: 'PUSH'
  },
  relay: {
    hasMore: jest.fn(() => true),
    loadMore: jest.fn(),
    isLoading: jest.fn(() => false)
  }
}

function loadSubject (props = {}, options = { disableLifecycleMethods: true }) {
  subject = shallow(<BillList {...defaultsDeep(props, defaults)} />, options)
}

// specs
afterEach(() => {
  subject = null
})

describe('#state', () => {
  it('enables animations by default', () => {
    loadSubject()
    expect(subject).toHaveState('disableAnimations', false)
  })

  it('disables animations on pop', () => {
    loadSubject({
      history: { action: 'POP' }
    })

    expect(subject).toHaveState('disableAnimations', true)
  })
})

describe('#componentDidMount', () => {
  it('re-enables animations on pop', () => {
    loadSubject({
      history: { action: 'POP' }
    }, {
      disableLifecycleMethods: false
    })

    expect(subject).toHaveState('disableAnimations', false)
  })
})

describe('#componentWillUnmount', () => {
  it('stores the visible bill count', () => {
    const bills = [{ id: 1 }, { id: 2 }, { id: 3 }]

    loadSubject({
      viewer: {
        bills: {
          edges: bills.map((node) => ({ node }))
        }
      }
    })

    subject.unmount()
    expect(session.get('last-search-count')).toEqual(`${bills.length}`)
  })
})

describe('#render', () => {
  it('shows the list', () => {
    loadSubject()
    expect(subject).toMatchSnapshot()
  })
})

describe('clicking load more', () => {
  it('loads another page', () => {
    loadSubject()
    subject.instance().didClickLoadMore()
    expect(defaults.relay.loadMore).toHaveBeenCalledWith(999, anything())
  })

  it('does nothing if there this is the last page', () => {
    defaults.relay.hasMore.mockReturnValue(false)
    loadSubject()
    subject.instance().didClickLoadMore()
    expect(defaults.relay.loadMore).not.toHaveBeenCalled()
  })

  it('does nothing when already loading', () => {
    defaults.relay.isLoading.mockReturnValue(true)
    loadSubject()
    subject.instance().didClickLoadMore()
    expect(defaults.relay.loadMore).not.toHaveBeenCalled()
  })
})
