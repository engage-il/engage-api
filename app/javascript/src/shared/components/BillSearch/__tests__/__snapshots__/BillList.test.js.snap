// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render shows the list 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-t7v3vx=""
  >
    <h2>
      Upcoming Bills
    </h2>
    <div>
      Jan 1st to Jan 7th
    </div>
    <div>
      Found 2 results.
    </div>
  </div>
  <TranslateAndFade
    disable={false}
  >
    <BillCell
      bill={
        Object {
          "id": 1,
        }
      }
      key="1"
    />
    <BillCell
      bill={
        Object {
          "id": 2,
        }
      }
      key="2"
    />
  </TranslateAndFade>
  <LoadMoreButton
    hasMore={true}
    onClick={[Function]}
    styles={
      Object {
        "data-css-1n5n5pb": "",
      }
    }
  />
</div>
`;
